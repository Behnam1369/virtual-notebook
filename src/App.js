import { Routes, Route, BrowserRouter } from 'react-router-dom';
import Editor from './components/Editor';
import style from './App.module.scss';
import Error404 from './components/Error404';

function App() {
  return (
    <div className={style.app}>
      <BrowserRouter>
        <Routes>
          <Route path="/pages/:id" element={<Editor />} />
          <Route path="/pages" element={<Editor />} />
          <Route path="/" element={<Editor />} />
          <Route path="*" element={<Error404 />} />
        </Routes>
      </BrowserRouter>
    </div>
  );
}

export default App;
