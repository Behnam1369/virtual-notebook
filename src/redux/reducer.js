import { configureStore } from '@reduxjs/toolkit';

// Page actions
const ADD_PAGE = 'ADD_PAGE';
const DELETE_PAGE = 'DELETE_PAGE';
const EDIT_PAGE = 'EDIT_PAGE';
const CANCEL_EDIT_PAGE = 'CANCEL_EDIT_PAGE';
const UPDATE_PAGE = 'UPDATE_PAGE';

// Element actions
const ADD_ELEMENT = 'ADD_ELEMENT';

const defaultState = {
  pages: [
    {
      id: 1,
      title: 'Home',
      elements: [
        { id: 1, style: 'h2', text: 'Front-end developer test project' },
        { id: 2, style: 'p', text: 'Your goal is to make a page that looks exactly like this one, and has the ability to create H1 text simply by typing / then 1, then typing text, and hitting enter.' },
      ],
    },
  ],
};

export default function reducer(state = defaultState, action) {
  switch (action.type) {
    case ADD_PAGE:
      return {
        ...state,
        pages: [...state.pages,
          {
            id: state.pages.length === 0 ? 1 : (state.pages[state.pages.length - 1].id) + 1,
            title: action.title,
            elements: [],
          }],
      };
    case DELETE_PAGE:
      return {
        ...state,
        pages: state.pages.filter((page) => page.id !== action.id),
      };
    case EDIT_PAGE:
      return {
        ...state,
        pages: state.pages.map((page) => (page.id === action.id
          ? { ...page, editing: true }
          : { ...page, editing: false })),
      };
    case UPDATE_PAGE:
      return {
        ...state,
        pages: state.pages.map((page) => (page.id === action.id
          ? { ...page, title: action.title, editing: false }
          : page)),
      };
    case CANCEL_EDIT_PAGE:
      return {
        ...state,
        pages: state.pages.map((page) => ({ ...page, editing: false })),
      };
    case ADD_ELEMENT:
      return {
        ...state,
        pages: state.pages.map((page) => (page.id === action.pageId
          ? {
            ...page,
            elements: [...page.elements, {
              id: page.elements.length === 0 ? 1 : (page.elements[page.elements.length - 1].id) + 1,
              style: action.style,
              text: action.text,
            }],
          }
          : page)),
      };
    default:
      return state;
  }
}

export function addPage(el) {
  return { ...el, type: ADD_PAGE };
}

export function deletePage(el) {
  return { ...el, type: DELETE_PAGE };
}

export function editPage(el) {
  return { ...el, type: EDIT_PAGE };
}

export function cancelEditPage(el) {
  return { ...el, type: CANCEL_EDIT_PAGE };
}

export function updatePage(el) {
  return { ...el, type: UPDATE_PAGE };
}

export function addElement(el) {
  return { ...el, type: ADD_ELEMENT };
}

const persistedState = localStorage.getItem('virtual_notebook')
  ? JSON.parse(localStorage.getItem('virtual_notebook'))
  : defaultState;

export const store = configureStore({ reducer, preloadedState: persistedState });
store.subscribe(() => {
  localStorage.setItem('virtual_notebook', JSON.stringify(store.getState()));
});
