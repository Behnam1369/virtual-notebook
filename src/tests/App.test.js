import userEvent from '@testing-library/user-event';
import TestRenderer from 'react-test-renderer';
import {
  render,
  screen,
  fireEvent,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';
import { store } from '../redux/reducer';
import App from '../App';

describe('Snapshot test:', () => {
  it('Rendered App component should match the snapshot', () => {
    const testRenderer = TestRenderer.create(
      <Provider store={store}>
        <App />
      </Provider>,
    ).toJSON();
    expect(testRenderer).toMatchSnapshot();
  });
});

describe('Integration tests for Pages:', () => {
  it('By default a home page should be available', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );
    fireEvent.click(screen.getByTestId('hamburgurButton'));
    expect(screen.getByTestId('pageList')).toHaveTextContent('Home');
  });

  it('Adding a page by dispatching an action', () => {
    store.dispatch({ type: 'ADD_PAGE', title: 'New Page' });
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    fireEvent.click(screen.getByTestId('hamburgurButton'));
    expect(screen.getByTestId('pageList')).toHaveTextContent('New Page');
  });

  it('Adding a page by submitting values on the form', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    fireEvent.click(screen.getByTestId('hamburgurButton'));
    const input = screen.getByTestId('newPageInput');
    fireEvent.change(input, { target: { value: 'Third Page' } });
    fireEvent.click(screen.getByText('Add Page'));
    expect(screen.getByTestId('pageList')).toHaveTextContent('Third Page');
  });

  it('Shouldn\'t be able to add a page without a title', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    fireEvent.click(screen.getByTestId('hamburgurButton'));
    const input = screen.getByTestId('newPageInput');
    fireEvent.change(input, { target: { value: '' } });
    fireEvent.click(screen.getByText('Add Page'));
    expect(screen.getByTestId('pageList').children).toHaveLength(3);
  });

  it('By clicking on the Edit button an input should be available for editing', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    fireEvent.click(screen.getByTestId('hamburgurButton'));
    const pageList = screen.getByTestId('pageList');
    const editButton = pageList.getElementsByTagName('button')[0];
    fireEvent.click(editButton);
    const input = screen.getByTestId('editingPageInput');
    expect(input).toBeInTheDocument();
  });

  it('By clicking on the Edit button an input should be available for editing', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    fireEvent.click(screen.getByTestId('hamburgurButton'));
    const pageList = screen.getByTestId('pageList');
    const input = screen.getByTestId('editingPageInput');
    fireEvent.change(input, { target: { value: 'New Home Page' } });
    const submitButton = pageList.getElementsByTagName('button')[0];
    fireEvent.click(submitButton);
    expect(pageList).toHaveTextContent('New Home Page');
  });

  it('By clicking on the delete button, it should remove a page', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    fireEvent.click(screen.getByTestId('hamburgurButton'));
    const pageList = screen.getByTestId('pageList');
    const deleteButton = pageList.getElementsByTagName('button')[5];
    fireEvent.click(deleteButton);
    expect(pageList.children).toHaveLength(2);
  });
});

describe('Integration tests for Elements:', () => {
  it('By default an input should be available', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const input = screen.getByTestId('newElementInput');
    expect(input).toBeInTheDocument();
  });

  it('Input should have a "p" class', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const input = screen.getByTestId('newElementInput');
    expect(input).toHaveClass('p');
  });

  it('By pressing "/" a selector should appear', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const input = screen.getByTestId('newElementInput');
    fireEvent.keyDown(input, { key: '/' });
    const selector = screen.getByTestId('selector');
    expect(selector).toBeInTheDocument();
  });

  it('Selector should contain an H1 option', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const input = screen.getByTestId('newElementInput');
    fireEvent.keyDown(input, { key: '/' });
    const selector = screen.getByTestId('selector');
    expect(selector).toHaveTextContent('H1');
  });

  it('By filtering "H1" Other options shouldn\' be available', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const input = screen.getByTestId('newElementInput');
    fireEvent.keyDown(input, { key: '/' });
    fireEvent.change(input, { target: { value: '/h1' } });
    const selector = screen.getByTestId('selector');
    expect(selector).toHaveTextContent('H1');
    expect(selector).not.toHaveTextContent('H2');
  });

  it('By Hitting enter an h1 markdown should be applied', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const input = screen.getByTestId('newElementInput');
    fireEvent.keyDown(input, { key: '/' });
    fireEvent.change(input, { target: { value: '/h1' } });
    userEvent.type(input, '{enter}');
    expect(input).toHaveClass('h1');
  });

  it('By Hitting one more time a P element with h1 class should be added', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>,
    );

    const input = screen.getByTestId('newElementInput');
    fireEvent.keyDown(input, { key: '/' });
    fireEvent.change(input, { target: { value: '/h1' } });
    userEvent.type(input, '{enter}');
    userEvent.type(input, '{enter}');
    const element = screen.getByTestId('editor').querySelector('p.h1');
    expect(element).toBeInTheDocument();
  });
});
