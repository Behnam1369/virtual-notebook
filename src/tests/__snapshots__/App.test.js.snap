// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test: Rendered App component should match the snapshot 1`] = `
<div
  className="app"
>
  <div
    className="container"
    data-testid="editor"
  >
    <div
      className="main"
    >
      <div
        className="header"
      >
        <div>
          <button
            data-testid="hamburgurButton"
            onClick={[Function]}
            type="button"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 15 15"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clipRule="evenodd"
                d="M1.5 3C1.22386 3 1 3.22386 1 3.5C1 3.77614 1.22386 4 1.5 4H13.5C13.7761 4 14 3.77614 14 3.5C14 3.22386 13.7761 3 13.5 3H1.5ZM1 7.5C1 7.22386 1.22386 7 1.5 7H13.5C13.7761 7 14 7.22386 14 7.5C14 7.77614 13.7761 8 13.5 8H1.5C1.22386 8 1 7.77614 1 7.5ZM1 11.5C1 11.2239 1.22386 11 1.5 11H13.5C13.7761 11 14 11.2239 14 11.5C14 11.7761 13.7761 12 13.5 12H1.5C1.22386 12 1 11.7761 1 11.5Z"
                fill="currentColor"
                fillRule="evenodd"
              />
            </svg>
          </button>
          <p>
            Main / Pages / Home
          </p>
        </div>
        <div>
          <p>
            <span>
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="11"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="11"
                />
                <path
                  d="M7 11V7a5 5 0 0 1 9.9-1"
                />
              </svg>
               
              Editing
            </span>
            |
            <span>
              Publish Space
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="6 9 12 15 18 9"
                />
              </svg>
            </span>
          </p>
        </div>
      </div>
      <div
        className="body"
      >
        <div
          className="options"
        >
          <div>
            <span>
              P
            </span>
            |
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 3.5a.5.5 0 0 0-1 0V9a.5.5 0 0 0 .252.434l3.5 2a.5.5 0 0 0 .496-.868L8 8.71V3.5z"
              />
              <path
                d="M8 16A8 8 0 1 0 8 0a8 8 0 0 0 0 16zm7-8A7 7 0 1 1 1 8a7 7 0 0 1 14 0z"
              />
            </svg>
            |
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M382.7 27.2c4.8 2.25 8.5 5.18 11.5 8.94-27.3 19.88-43.5 52.15-53.1 84.86-11.6 39.4-14.1 80.2-14.1 104.8h18c0-23.4 2.5-62.9 13.3-99.7 8.6-29.19 22.3-56.12 43.4-73.03 1.2 4.94 1.9 10.46 2.1 16.52.6 22.68-5.6 52.11-14.4 81.61-8.9 29.6-20.3 59.4-29.6 83.9-3.8 10.2-14.7 50.7-11.4 54.7 15.9 21.8 26.6 48 26.6 76v3.7l2.6 2.7c14.3 14.2 15.9 25.5 11.7 37.6-4.3 12.1-16.3 25.3-32.5 36.6-32.3 22.8-80.8 38.4-100.8 38.4s-68.5-15.6-100.8-38.4c-16.2-11.3-28.2-24.5-32.5-36.6-4.2-12.1-2.6-23.4 11.7-37.6l2.6-2.7v-3.7c0-28 10.7-54.2 26.6-76 3.3-4-7.6-44.5-11.4-54.7-9.3-24.5-20.7-54.3-29.6-83.9-8.8-29.5-15-58.89-14.4-81.59.2-6.1.9-11.6 2.1-16.5 21.1 16.9 34.8 43.8 43.4 72.99 10.8 36.8 13.3 76.3 13.3 99.7h18c0-24.6-2.5-65.4-14.1-104.8-9.6-32.7-25.8-64.99-53.1-84.89 3-3.7 6.7-6.7 11.5-8.9 3.6.9 13.2 6.1 23.4 16.9 11.2 11.9 23.7 29.4 34.5 50.1 21.7 41.39 37 95.79 27.9 145.99 12.4-3.9 26.7-5.9 40.9-5.9s28.5 2 40.9 5.9c-9.1-50.2 6.2-104.6 27.9-146 10.8-20.72 23.3-38.19 34.5-50.13 10.2-10.8 19.8-15.94 23.4-16.87zM304 308.8c-13.7 0-25 11.3-25 25s11.3 25 25 25 25-11.3 25-25-11.3-25-25-25zm-96 0c-13.7 0-25 11.3-25 25s11.3 25 25 25 25-11.3 25-25-11.3-25-25-25zm96 18c4 0 7 3 7 7s-3 7-7 7-7-3-7-7 3-7 7-7zm-96 0c4 0 7 3 7 7s-3 7-7 7-7-3-7-7 3-7 7-7zm48 34c-8.6 0-17 1-24.4 3.5-3.7 1.2-7.1 2.7-10.2 5.3-3.2 2.5-6 6.7-6.4 11.6l-.4 6 32.4 16.2v27c-.3.2-2.2.7-5.4 0-6.5-1.6-17.1-8-26.1-21.6l-15 10c11 16.6 24.4 26.2 36.9 29.1 6.3 1.5 12.3 3.4 18.6 3.4 6.3 0 12.3-1.9 18.6-3.4 12.5-2.9 25.9-12.5 36.9-29.1l-15-10c-9 13.6-19.6 20-26.1 21.6-3.2.7-5.1.2-5.4 0v-27l32.4-16.2-.4-6c-.4-4.9-3.2-9.1-6.4-11.6-3.1-2.6-6.5-4.1-10.2-5.3-7.4-2.5-15.8-3.5-24.4-3.5zm0 18c5.3 0 10.3.7 14.6 1.7l-14.6 7.2-14.6-7.2c4.3-1 9.3-1.7 14.6-1.7z"
              />
            </svg>
            |
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="17"
                x2="7"
                y1="7"
                y2="17"
              />
              <polyline
                points="17 17 7 17 7 7"
              />
            </svg>
            0
          </div>
          <div>
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
              />
              <polyline
                points="22 4 12 14.01 9 11.01"
              />
            </svg>
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 10h-1.26A8 8 0 1 0 9 20h9a5 5 0 0 0 0-10z"
              />
            </svg>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 12c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
          </div>
        </div>
        <h1>
          Home
        </h1>
        <p
          className="h2"
          onClick={[Function]}
        >
          Front-end developer test project
        </p>
        <p
          className="p"
          onClick={[Function]}
        >
          Your goal is to make a page that looks exactly like this one, and has the ability to create H1 text simply by typing / then 1, then typing text, and hitting enter.
        </p>
        <input
          className="input p"
          data-testid="newElementInput"
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Type / for blocks"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
